File: part2/protocols/bin_mem.py
# Apache Thrift Binary Protocol in Python

from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

trans = TTransport.TMemoryBuffer()
proto = TBinaryProtocol.TBinaryProtocol(trans)
proto.writeString("Hello Thrift Serialization")
print ("Wrote %d bytes to the TMemoryBuffer" % (trans.cstringio_buf.tell()))

trans.cstringio_buf.seek(0)
msg = proto.readString()
print("Recovered string: %s" % (msg))

File: part2/protocols/bin_mem.cpp
// Apache Thrift Binary Protocol in C++ 

#include <thrift/transport/TBufferTransports.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <memory>
#include <iostream>
#include <string>

using namespace apache::thrift::transport;
using namespace apache::thrift::protocol;
using std::make_shared;

int main()
{
    auto trans = make_shared<TMemoryBuffer>(4096);
    TBinaryProtocol proto(trans);

    int i = proto.writeString(std::string("Hello Thrift Serialization"));
    std::cout << "Wrote " << i << " bytes to the TMemoryBuffer" << std::endl;

    std::string msg;
    i = proto.readString(msg);
    std::cout << "Read " << i << " bytes from the TMemoryBuffer" << std::endl;
    std::cout << "Recovered string: " << msg << std::endl;
}

